### PR TITLE
AEA-456 Keep key ordering when loading/dumping configuration.

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

Keep key ordering when loading or dumping a configuration.

## Fixes

- #932

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/configurations/base.py [hunk 1]
```diff
@@ -812,22 +861,26 @@ def package_dependencies(self) -> Set[ComponentId]:
...
-            "dependencies": self.dependencies,
-            "description": self.description,
-            "config": self.config,
-        }
+        return OrderedDict(
+            {
+                "name": self.name,
+                "author": self.author,
+                "version": self.version,
+                "description": self.description,
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:869]: I think general order should be common headers (name ...fingerprint ignore patterns), framework package dependencies, module specifics, python package dependencies. Thoughts?

File: aea/configurations/base.py [hunk 2]
```diff
@@ -812,22 +861,26 @@ def package_dependencies(self) -> Set[ComponentId]:
...
+                "license": self.license,
+                "aea_version": self.aea_version,
+                "fingerprint": self.fingerprint,
+                "fingerprint_ignore_patterns": self.fingerprint_ignore_patterns,
+                "class_name": self.class_name,
+                "protocols": sorted(map(str, self.protocols)),
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:875]: In the spirit of the above, I would move protocols above class_name, I would also have class_name, 
 config, excluded_protocols, restricted_to_protocols in this order. I'd have dependencies last.

File: aea/configurations/base.py [hunk 3]
```diff
@@ -1014,22 +1069,24 @@ def package_dependencies(self) -> Set[ComponentId]:
...
+            {
+                "name": self.name,
+                "author": self.author,
+                "version": self.version,
+                "description": self.description,
+                "license": self.license,
+                "aea_version": self.aea_version,
+                "fingerprint": self.fingerprint,
+                "fingerprint_ignore_patterns": self.fingerprint_ignore_patterns,
+                "protocols": sorted(map(str, self.protocols)),
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:1082]: In spirit of above, swap contracts and protocols order, then "behaviours", "handlers", "models", "dependencies"

File: aea/configurations/base.py [hunk 4]
```diff
@@ -1203,26 +1260,28 @@ def default_ledger(self, ledger_id: str):
...
+        return OrderedDict(
+            {
+                "agent_name": self.agent_name,
+                "author": self.author,
+                "version": self.version,
+                "description": self.description,
+                "license": self.license,
+                "aea_version": self.aea_version,
+                "fingerprint": self.fingerprint,
+                "fingerprint_ignore_patterns": self.fingerprint_ignore_patterns,
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:1272]: In spirit of above, would have: "connections", "contracts", "protocols", "skills", "default_connection", "default_ledger", "ledger_apis", "logging_config", "private_key_paths", "registry_path"

File: aea/configurations/base.py [hunk 5]
```diff
@@ -1447,19 +1510,21 @@ def component_type(self) -> ComponentType:
...
+            {
+                "name": self.name,
+                "author": self.author,
+                "version": self.version,
+                "description": self.description,
+                "license": self.license,
+                "aea_version": self.aea_version,
+                "fingerprint": self.fingerprint,
+                "fingerprint_ignore_patterns": self.fingerprint_ignore_patterns,
+                "dependencies": self.dependencies,
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:1523]: In spirit of above, would have, "class_name", "path_to_contract_interface", "dependencies"

File: tests/test_cli/test_remove/test_skill.py [hunk 6]
```diff
@@ -69,7 +69,7 @@ def setup_class(cls):
         # change default registry path
         config = AgentConfig.from_json(yaml.safe_load(open(DEFAULT_AEA_CONFIG_FILE)))
         config.registry_path = os.path.join(ROOT_DIR, "packages")
-        yaml.safe_dump(config.json, open(DEFAULT_AEA_CONFIG_FILE, "w"))
+        yaml.safe_dump(dict(config.json), open(DEFAULT_AEA_CONFIG_FILE, "w"))
```
Comment (DavidMinarsch) [inline tests/test_cli/test_remove/test_skill.py:72]: why not "dump"?

File: tests/test_cli/test_add/test_skill.py [hunk 7]
```diff
@@ -374,7 +374,7 @@ def setup_class(cls):
         # change default registry path
         config = AgentConfig.from_json(yaml.safe_load(open(DEFAULT_AEA_CONFIG_FILE)))
         config.registry_path = os.path.join(ROOT_DIR, "packages")
-        yaml.safe_dump(config.json, open(DEFAULT_AEA_CONFIG_FILE, "w"))
+        yaml.safe_dump(dict(config.json), open(DEFAULT_AEA_CONFIG_FILE, "w"))
```
Comment (DavidMinarsch) [inline tests/test_cli/test_add/test_skill.py:377]: why not "dump"?

File: contract.yaml [hunk 8]
```diff
@@ -0,0 +1,13 @@
+aea_version: '>=0.2.4'
+author: fetchai
+class_name: ''
```
Comment (DavidMinarsch) [inline contract.yaml:3]: remove!

File: protocol.yaml [hunk 9]
```diff
@@ -0,0 +1,13 @@
+aea_version: '>=0.2.4'
+author: fetchai
+class_name: ''
+dependencies: {}
```
Comment (DavidMinarsch) [inline protocol.yaml:4]: remove!

File: tests/test_cli/test_add/test_skill.py (hunk 7 above)
Comment (marcofavorito) [reply tests/test_cli/test_add/test_skill.py:377]: you mean why not just `yaml.dump`?

File: aea/configurations/base.py (hunk 1 above)
Comment (marcofavorito) [reply aea/configurations/base.py:869]: yes, definitely. That includes some refactoring. We should move common header fields in `super().json` and then update the dictionary in the children classes.

**Pull request merged**